README.md
# Device-Camera-Go

This repository contains a camera device service compatible with EdgeX Foundry. The project utilizes [EdgeX Device SDK Go](https://github.com/edgexfoundry/device-sdk-go) to communicate with dependent EdgeX Go services. This means the service relies on the appropriate EdgeX services to be running in order to start up and function properly (Consul, Core Data, Core Metadata and others).

Use Device-Camera-Go to manage interactions between your EdgeX applications and IP network cameras. This includes the ability to capture still images or RTP video streams from ONVIF compliant IP cameras connected to your network. It also supports accessing Axis cameras, making available a similar but different interface to perform functions with the camera.

The project is intended to provide both functionality to serve as a springboard for visual application development at the edge, as well as a reference example of how to implement device services using the EdgeX Device SDK Go package.

This service has been tested with the following configuration:

IP cameras:

- 3x AXIS M3046-V Network Camera with firmware version: 6.15.6
- 1x HIKVISION Camera with firmware version: V5.4.1 build 160525

Development System:

- Oracle VirtualBox VM running Ubuntu 16.04 with 2 cores, 6GB RAM and 256MB video memory.

# Getting Started

The instructions below are the minimal steps needed to get you up and going quickly.

Additional details and answers related to EdgeX itself are available on:

- [EdgeX Wiki](https://docs.edgexfoundry.org/Ch-GettingStartedUsers.html)

# Prerequisites - Developer Setup

Perform the following steps to set up your development system. These are necessary to build and extend the device-camera-go component.

Further detailed information is available on [EdgeX-Go repository](https://github.com/edgexfoundry/edgex-go/blob/master/docs/getting-started/Ch-GettingStartedGoDevelopers.rst).

1. Install **Go 1.10** and Glide (package manager) using instructions found here:

   - https://github.com/golang/go/wiki/Ubuntu

   - NOTE: If upgrading from an earlier version of Go, it is best to first remove it; ref: https://golang.org/doc/install

   - Install **dep** with:

     ```
     $ curl https://raw.githubusercontent.com/golang/dep/master/install.sh | sh
     ```

2. Install **Docker**.

   - If upgrading from an earlier version of Docker, BKM is to first remove it using:

     ```
     $ sudo apt-get remove docker docker-engine docker.io
     ```

   - Follow instructions from [Install Docker using repository](https://docs.docker.com/install/linux/docker-ce/ubuntu/#install-using-the-repository)

3. Install **Dep**.
   ```
   $ curl https://raw.githubusercontent.com/golang/dep/master/install.sh | sh
   ```

4. (optional) Install an **IDE** as appropriate. **[JetBrains Go Land](https://www.jetbrains.com/go/)** is a popular choice, alternate options below.

   ```
   $ tar -xzf goland-2018.2.1.tar.gz
   $ ./bin/goland.sh
   ```

   - Alternate: [Visual Studio](https://code.visualstudio.com/)
   - Alternate: [Sublime Text](https://www.alexedwards.net/blog/streamline-your-sublime-text-and-go-workflow)
   - ...

# Preparing EdgeX Stack

1. The easiest way to launch and explore device-camera-go using existing EdgeX services is to clone the edgexfoundry/developer-scripts repository and run the published Docker containers for the **EdgeX Delhi 0.7.1 release**.

   ```
   $ cd ~/dev
   $ git clone https://github.com/edgexfoundry/developer-scripts.git
   $ cd ~/dev/developer-scripts/compose-files
   ```

   At time of writing, you will see that the latest docker-compose.yml is equivalent to Delhi 0.7.1.

2. The device-camera-go device service does not require other EdgeX device services. To reduce runtime footprint a bit, and possible confusion about new devices, comment out the device-virtual section (using # at beginning of the line) in **./compose-files/docker-compose.yml**.

   ```
   # device-virtual:
   #   image: edgexfoundry/docker-device-virtual:0.6.0
   #   ports:device-virtual:
   ...
   ```

3. If you are already running MongoDB on its default port (27017), you will want to also update **./compose-files/docker-compose.yml** to assign a unique port for EdgeX MongoDB (e.g., 27018 as the port referenced by your local system):

   ```
   mongo:
     ...
     ports:
      - "27018:27017"
   ```

4. The first time launching EdgeX, Docker will automatically **pull the EdgeX images** to your system. This requires an Internet connection and allowance through proxies/firewall. Detailed information about pulling EdgeX containers is available here: https://docs.edgexfoundry.org/Ch-GettingStartedUsersNexus.html

# Preparing Device-Camera-Go

1. Get latest device-camera-go source code:

   ```
   cd $GOPATH/src/github.com/edgexfoundry-holding
   git clone https://github.com/edgexfoundry-holding/device-camera-go.git
   ```

   Note: You may alternately clone to a separate project folder so long as you create a symlink to it from $GOPATH/src/github.com/edgexfoundry-holding folder.

2. Build a docker image by executing the following.
   ```
   cd ./device-camera-go
   make docker
   ```

   Note: If you require a *proxy* to access the Internet, confirm the parameters to the docker build command (found in Makefile).

3. To build within an IDE such as GoLand, load device-camera-go project, execute the following within View/Tools/Terminal:

   ```
   $ make prepare
   $ make build
   ```

   Alternately, invoke these same commands in a terminal from the root folder of the project.

   NOTE: make prepare will error if the manifest and lock files already exist (named Gopkg.toml and Gopkg.lock respectively). In this case run **make update** instead.

4. Run the following instructions to patch device-sdk-go (needed for Delhi branch only):

   ```
   sed -i '71i // GetDeviceByName returns device if it exists in EdgeX registration cache.\nfunc (s *Service) GetDeviceByName(name string) (models.Device, error) {\n   device, ok := cache.Devices().ForName(name)\n   if !ok {\n      msg := fmt.Sprintf("Device %s cannot be found in cache", name)\n      common.LoggingClient.Info(msg)\n      return models.Device{}, fmt.Errorf(msg)\n   }\n   return device, nil\n}' vendor/github.com/edgexfoundry/device-sdk-go/manageddevices.go

# Running device-camera-go

Ready-set-go!

0. Add the device-camera-go:develop image to the bottom of your docker-compose.yml file.
```
  device-camera-go:
    image: device-camera-go:develop
    container_name: device-camera-go
    hostname: device-camera-go
    network_mode: host
    depends_on:
      - metadata
```

1. **Launch the EdgeX stack** by instructing Docker to use developer-scripts/compose-files/docker-compose.yml to run the services in detached state. This way control will return to your terminal window. Of course you can leave off the -d parameter and open additional terminals if you prefer watching EdgeX-specific logging activities in real time.

   ```
   $ cd ~/dev/developer-scripts/compose-files
   $ docker-compose up -d
   ```

2. **Confirm EdgeX services** are running properly by navigating to the Consul dashboard in your browser:

   - http://127.0.0.1:8500/ui/#/dc1/services

     [image of green/passing services]

3. **Modify your /etc/hosts file** to override DNS resolution to resolve EdgeX service URLs. Add a single line below your existing localhost entry:

   > 127.0.0.1	localhost
   > 127.0.0.1	edgex-core-command edgex-core-metadata edgex-core-data
   >
   > ...

   Alternately, adjust EdgeX service URLs used in this guide and the tests/postman import file so they point to  "127.0.0.1", "localhost".

4. **Modify the ./run.sh script** with appropriate options (described below) and then launch the device-camera-go service.
   NOTE: Be sure to update the parameters you supply to device-camera-go with expected sources and IP address ranges according to your network topology and IP camera deployment and configuration:

   ```
   $ cd $GOPATH/src/github.com/edgexfoundry-holding/device-camera-go
   $ make run
   $ ./run
   ```

# Device-Camera-Go Configuration

Device-Camera-Go accepts the following runtime parameters.

| Parameter          | Default                    | Description                                                  |
| ------------------ | -------------------------- | ------------------------------------------------------------ |
| -registry / -r     | false                      | If this flag is set, the device-camera-go service is registered with and monitored by Consul |
| -confdir / -c      | ./res                      | Allows you to specify an alternate configuration directory. The path you specify after this parameter will be used to locate the configuration.toml and related files for loading your specified deploy-time set of camera devices. |
| -ip                | 192.168.0.1-30             | IP address/range for nmap scan.<br />                        |
| -mask              | ""                         | Network mask for nmap scan<br />Subnet scan example:<br/>-ip "10.25.35.*" - mask "/24"<br /><br />Single IP example:<br />-ip = "10.1.24.101" -mask = "32" -scanduration = "10s"<br /> |
| -scanduration      | "15s"                      | Controls how long to allow IP groups to be scanned before timing out. The value may specify units of seconds, minutes, or hours ("s", "m", "h", respectively) |
| -interval          | 60                         | Frequency of scans to schedule. Units are in seconds. The lower the number the more responsive, but greater network overhead and system I/O required. At a minimum, the value must be 10 seconds longer than the configured scan duration but for practical purposes should be set to much larger values to allow IP groups to scan (highly relative to # IP and # ports being scanned). |
| -source  ...       | (see description)          | Source to scan. Provides a way to segment whether ONVIF and/or Axis cameras should be discovered. When no colon/port is provided, default port is 554 for Axis and 80 for ONVIF.<br />Examples:<br />-source axis<br />-source onvif<br />-source axis -source onvif<br />-source axis:554 -source onvif:80<br />-source axis:554,557,558 -source onvif:80,8000,8020 |
| -tagsFile          | ./res/tags.json            | Path to file containing camera tags. These are used on startup and will reflect any changes occurring by invoking /tags command. |
| -cameracredentials | ./res/cam-credentials.conf | Location and name of the file where camera access credentials Path to file containing credentials for ONVIF cameras. This holds a user and password, tab-separated that is configured on the camera for purpose of being accessed by the device-camera-go service. |

```
docker run --name device-camera-go -it --network host --rm hub.docker.com/device-camera-go:v0.1.0 <parameters>
```

```
$ go run main.go -registry -source onvif -source axis
```

## Build - and other Make Targets

| Makefile Command | Action Performed                                             |
| ---------------- | ------------------------------------------------------------ |
| make prepare     | Initializes the project manifest and dependencies            |
| make update      | Updates components                                           |
| make build       | Builds camera-device-go binary                               |
| make run         | Converts run.sh content into ./run launcher                  |
| make test        | Invokes gometalinter (after installing for you if needed) and go test |
| make clean       | Removes camera-device-go binary                              |

## Tests

A set of handy Postman links are available in the **./tests** subfolder.

Since name prefixes are prepended to camera serialnumbers to uniquely identify (and namespace) your EdgeX IP camera devices. This constructs an immutable way to access a discovered IP camera; the name will not change across service start/stop or even removing information from the backend database. As a result, you will find it useful to add Postman requests using these device names to verify results.

For example, replace the serial number ACCC8E8439F0 with your camera's serial number on this URI:

- http://edgex-core-command:48082/api/v1/device/name/edgex-camera-onvif-ACCC8E8439F0

## Persisted Resources

File-based resources are kept in the **./res** folder. This can be configured with the -confdir parameter to point to a different location.  and cache backing stores and stored in ./res folder.

## Configuration.toml

The ./res/configuration.toml will dynamically define and register devices based on the information provided. It links to the described EdgeX device profile (constructing virtual Axis camera devices) as a template to populate with these static cameras known at deploy-time.

## Tags File

An example of values to populate camera tags is found in the ./res folder. Note that the keys are simply the device serial number. So when the associated device is discovered using both ONVIF and Axis, you will find a separate device name registered with EdgeX for each device profile (uniquely named per a name prefix).

> {

CHANGELOG.md

<a name="EdgeX Camera Device Service (found in device-camera-go) Changelog"></a>
## EdgeX Camera Device Service
[Github repository](https://github.com/edgexfoundry/device-camera-go)

### Change Logs for EdgeX Dependencies
- [device-sdk-go](https://github.com/edgexfoundry/device-sdk-go/blob/main/CHANGELOG.md)
- [go-mod-core-contracts](https://github.com/edgexfoundry/go-mod-core-contracts/blob/main/CHANGELOG.md)
- [go-mod-bootstrap](https://github.com/edgexfoundry/go-mod-bootstrap/blob/main/CHANGELOG.md)
- [go-mod-messaging](https://github.com/edgexfoundry/go-mod-messaging/blob/main/CHANGELOG.md) (indirect dependency)
- [go-mod-registry](https://github.com/edgexfoundry/go-mod-registry/blob/main/CHANGELOG.md)  (indirect dependency)
- [go-mod-secrets](https://github.com/edgexfoundry/go-mod-secrets/blob/main/CHANGELOG.md) (indirect dependency)
- [go-mod-configuration](https://github.com/edgexfoundry/go-mod-configuration/blob/main/CHANGELOG.md) (indirect dependency)

## [v2.2.0] Kamakura - 2022-05-11  (Not Compatible with 1.x releases)

### Features ✨
- Enable security hardening ([#583c54f](https://github.com/edgexfoundry/device-camera-go/commits/583c54f))

### Bug Fixes 🐛
- **snap:** Expose parent directory in device-config plug ([#f6f2141](https://github.com/edgexfoundry/device-camera-go/commits/f6f2141))

### Code Refactoring ♻
- **snap:** Remove redundant content indentifier ([#a81407d](https://github.com/edgexfoundry/device-camera-go/commits/a81407d))

### Documentation 📖
- **snap:** remove two environment configuration overrides ([#ccfe6b2](https://github.com/edgexfoundry/device-camera-go/commits/ccfe6b2))

### Build 👷
- Update to latest SDK w/o ZMQ on windows ([#9b3fa78](https://github.com/edgexfoundry/device-camera-go/commits/9b3fa78))
    ```
    BREAKING CHANGE:
    ZeroMQ no longer supported on native Windows for EdgeX
    MessageBus
    ```
- **snap:** source metadata from central repo ([#193](https://github.com/edgexfoundry/device-camera-go/issues/193)) ([#3f3ab22](https://github.com/edgexfoundry/device-camera-go/commits/3f3ab22))

### Continuous Integration 🔄
- gomod changes related for Go 1.17 ([#82a10a5](https://github.com/edgexfoundry/device-camera-go/commits/82a10a5))
- Go 1.17 related changes ([#569337f](https://github.com/edgexfoundry/device-camera-go/commits/569337f))

## [v2.1.0] Jakarta - 2021-11-18  (Not Compatible with 1.x releases)

### Features ✨
- Update config files to include secretsfile ([#94ad87a](https://github.com/edgexfoundry/device-camera-go/commits/94ad87a))

### Documentation 📖
- Add comment for secretsfile ([#6f6b908](https://github.com/edgexfoundry/device-camera-go/commits/6f6b908))

### Build 👷
- Update to released SDK and go-mods ([#4ddf907](https://github.com/edgexfoundry/device-camera-go/commits/4ddf907))
- Bump to device-sdk-go[@v2](:/v2).0.1-dev.23 and go-mod-core-contracts[@v2](:/v2).0.1-dev.26 ([#89ba74a](https://github.com/edgexfoundry/device-camera-go/commits/89ba74a))
- **snap:** upgrade base to core20 ([#ee1913c](https://github.com/edgexfoundry/device-camera-go/commits/ee1913c))
- **snap:** bump edgex-snap-hooks to support secretsfile config ([#e3dffcc](https://github.com/edgexfoundry/device-camera-go/commits/e3dffcc))

## [v2.0.1] Ireland - 2021-10-08  (Not Compatible with 1.x releases)
- Use correct constants for authentication ([#4e5edb8](https://github.com/edgexfoundry/device-camera-go/commits/4e5edb8))
- Update PascalCase for device resource names ([#138](https://github.com/edgexfoundry/device-camera-go/issues/138)) - Aligns code with device profile changes made in PR 111 ([#8e9dd4c](https://github.com/edgexfoundry/device-camera-go/commits/8e9dd4c))
- Update all TOML to use quote and not single-quote ([#d3b9f4b](https://github.com/edgexfoundry/device-camera-go/commits/d3b9f4b))

## [v2.0.0] Ireland - 2021-06-30  (Not Compatible with 1.x releases)

### Features ✨
- Enable using MessageBus as the default ([#9d8893d](https://github.com/edgexfoundry/device-camera-go/commits/9d8893d))
- Remove Logging configuration ([#2f6e8f2](https://github.com/edgexfoundry/device-camera-go/commits/2f6e8f2))
- **security:** Get Camera creds from SecretProvider - Add implementation of getting camera credentials from secret provider / secret store instead of configuration toml directly - Remove plain text camera credentials from toml config file - Refactor camera credentials to allow multiple different camera credentials using DeviceList instead of Driver config ([#0fadece](https://github.com/edgexfoundry/device-camera-go/commits/0fadece))

### Style
- Use pascal case in the sample Profiles ([#b9d8927](https://github.com/edgexfoundry/device-camera-go/commits/b9d8927))

### Code Refactoring ♻
- remove unimplemented InitCmd/RemoveCmd configuration ([#d1479f8](https://github.com/edgexfoundry/device-camera-go/commits/d1479f8))
- Remove the dependency of pkg/errors ([#1083b87](https://github.com/edgexfoundry/device-camera-go/commits/1083b87))
- use v2 SDK ([#3339319](https://github.com/edgexfoundry/device-camera-go/commits/3339319))
- Update to assign and uses new Port Assignments and service key ([#fe94c2d](https://github.com/edgexfoundry/device-camera-go/commits/fe94c2d))
    ```
    BREAKING CHANGE:
    Device Camera default port number has changed to 59985
    ```
    ```
    BREAKING CHANGE:
    Device Camera service key has changed to device-camera
    ```
### Documentation 📖
- update README ([#acb6ef3](https://github.com/edgexfoundry/device-camera-go/commits/acb6ef3))
- Add badges to readme ([#1610273](https://github.com/edgexfoundry/device-camera-go/commits/1610273))
### Build 👷
- update snap build ([#c46a067](https://github.com/edgexfoundry/device-camera-go/commits/c46a067))
- update build files for v2 ([#d808c09](https://github.com/edgexfoundry/device-camera-go/commits/d808c09))
- **snap:** update epoch and release version to Ireland ([#95a98a1](https://github.com/edgexfoundry/device-camera-go/commits/95a98a1))

provider/edgexservice.go
// -*- mode: Go; indent-tabs-mode: t -*-
//
// Copyright (C) 2018-2019 Intel Corporation
//
// SPDX-License-Identifier: Apache-2.0

// Package cameradiscoveryprovider implements a CameraDiscovery provider which scans
// network for existence of compatible/requested IP cameras and registers these
// as EdgeX devices, with corresponding management/control interfaces.
package cameradiscoveryprovider

// Implements EdgeX ProtocolDriver interface
import (
	"fmt"
	"strings"
	"time"

	"github.com/edgexfoundry/device-sdk-go"
	ds_models "github.com/edgexfoundry/device-sdk-go/pkg/models"
	"github.com/edgexfoundry/go-mod-core-contracts/clients/logger"
	contract "github.com/edgexfoundry/go-mod-core-contracts/models"
)

const (
	// CameraManagementProfileName currently reuses Axis device profile to create a mock Camera Management Provider
	// as a placeholder EdgeX device (i.e., for the device service itself).
	CameraManagementProfileName string = "camera-profile-axis"
)

// AppLoggingClient overrides the implementation of the EdgeX logger.LoggingClient interface.
// It delegates logging to centralize any errors thrown by act of logging itself.
// It will always return nil or panic.
type AppLoggingClient interface {
	SetLogLevel(logLevel string) error
	Debug(msg string, labels ...string)
	Error(msg string, labels ...string)
	Info(msg string, labels ...string)
	Trace(msg string, labels ...string)
	Warn(msg string, labels ...string)
}
type appLogger struct {
	lc        logger.LoggingClient
	msgPrefix string
	count     int
	thresh    int
}

func newAppLogger(lc logger.LoggingClient) AppLoggingClient {
	return &appLogger{lc: lc, count: 0, thresh: 0, msgPrefix: ""}
	// Prepend any value to this service's logs using msgPrefix.
	// Example:
	//return &appLogger{lc: lc, count: 0, thresh: 0, msgPrefix: fmt.Sprintf("[%s] - ", device.RunningService().Name())}
}

func (l *appLogger) SetLogLevel(level string) error {
	return l.lc.SetLogLevel(level)
}
func (l *appLogger) Error(msg string, labels ...string) {
	new := make([]interface{}, len(labels))
	for i, v := range labels {
		new[i] = v
	}
	l.lc.Error(l.msgPrefix+msg, new...)
}

func (l *appLogger) Warn(msg string, labels ...string) {
	new := make([]interface{}, len(labels))
	for i, v := range labels {
		new[i] = v
	}
	l.lc.Warn(l.msgPrefix+msg, new...)
}

func (l *appLogger) Info(msg string, labels ...string) {
	new := make([]interface{}, len(labels))
	for i, v := range labels {
		new[i] = v
	}
	l.lc.Info(l.msgPrefix+msg, new...)
}

func (l *appLogger) Debug(msg string, labels ...string) {
	new := make([]interface{}, len(labels))
	for i, v := range labels {
		new[i] = v
	}
	l.lc.Debug(l.msgPrefix+msg, new...)
}

func (l *appLogger) Trace(msg string, labels ...string) {
	new := make([]interface{}, len(labels))
	for i, v := range labels {
		new[i] = v
	}
	l.lc.Trace(l.msgPrefix+msg, new...)
}

// AppCache holds elements related to application layer's device service caches.
// Currently segmented by tags and camera device types (i.e., a separate cache per device profile).
type AppCache struct {
	CamInfoCache  *CamInfo // In Memory device cache for discovered cameras
	InfoFileOnvif string   // File to back onvif device cache
	InfoFileAxis  string   // File to back axis device cache
	TagCache      *Tags    // Tags to be attached to cameras
	TagsFile      string   // File to initialize TagCache
}

// CameraDiscoveryProvider holds service level objects
type CameraDiscoveryProvider struct {
	lc                 AppLoggingClient
	asyncChan          chan<- *ds_models.AsyncValues
	options            *Options
	intervalTicker     *time.Ticker
	scanDurationTicker *time.Ticker
	ac                 *AppCache
}

// New instantiates CameraDiscoveryProvider
func New(options *Options, ac *AppCache) *CameraDiscoveryProvider {
	var p CameraDiscoveryProvider
	p.options = options
	p.ac = ac
	return &p
}

// DisconnectDevice is called by the SDK for protocol specific disconnection from device service.
func (p *CameraDiscoveryProvider) DisconnectDevice(deviceName string, protocols map[string]contract.ProtocolProperties) error {
	p.lc.Warn(fmt.Sprintf("DisconnectDevice CALLED: We can set state of devices, and update CoreMetadata..."))
	return nil
}

// Initialize performs protocol-specific initialization for the device
// service. The given *AsyncValues channel can be used to push asynchronous
// events and readings to EdgeX Core Data.
func (p *CameraDiscoveryProvider) Initialize(lc logger.LoggingClient, asyncCh chan<- *ds_models.AsyncValues) error {
	p.lc = newAppLogger(lc)
	p.asyncChan = asyncCh
	p.lc.Trace(fmt.Sprintf("CameraDiscoveryProvider Initialize called with options: %v", p.options))
	// ==============================
	// Validate and normalize inputs
	// ==============================
	// ScanDuration and Interval
	duration, err := time.ParseDuration(p.options.ScanDuration)
	if err != nil {
		p.lc.Error(fmt.Sprintf("Invalid ScanDuration. See help for examples."))
		return err
	}
	minWaitSeconds := 10
	if p.options.Interval <= int(duration.Seconds())+minWaitSeconds {
		err = fmt.Errorf("Must provide more than %d seconds between discovery scans!  Interval[%d] > ScanDuration[%v]", minWaitSeconds, p.options.Interval, duration.Seconds())
		return err
	}
	// IP and Mask - prepend '/'' to netmask if any value is specified
	if len(p.options.NetMask) > 0 {
		if p.options.NetMask[0] != '/' {
			p.options.NetMask = "/" + p.options.NetMask
		}
	}
	// PortList to scan; fed from supplied SourceFlags parameter(s)
	var portList string
	portList, err = p.buildPortList(*p.options)
	if err != nil {
		p.lc.Error(err.Error())
		return err
	}
	// ==============================
	// Load Camera Info cache(s)
	// ==============================
	err = p.loadCameraInfoCaches()
	if err != nil {
		p.lc.Warn(err.Error())
		// Existence of a camera info cache is not mandatory, continue with Initialize
		err = nil
	}
	// ==============================
	// Load Camera Tags cache
	// ==============================
	err = p.ac.TagCache.LoadTags(p.ac.TagsFile)
	if err != nil {
		p.lc.Warn(err.Error())
		// Existence of a tag cache is not mandatory, continue with Initialize
		err = nil
	}

	// ==============================
	// Load Credentials from Vault
	// ==============================
	// Perform our equivalent of a configuration registry lookup during service initialization.
	// This is performed against the res/configuration.toml using logic that currently matches
	// internal DSDK behavior. It uses values of a couple constants also defined there.
	credConfig, err := p.LoadCredFromFile(p.options.ConfProfile, p.options.ConfDir, p.options.CredentialsFile)
	if (err != nil) {
		p.lc.Error(err.Error())
	}
	p.options.Credentials = credConfig.Credentials
	// NOTE: We are now finished with cameraCredentialsFile.


	// ==============================
	// Report Info to logs for troubleshooting
	// ==============================
	p.lc.Info(fmt.Sprintf("Service Initialization:"))
	p.lc.Info(fmt.Sprintf("SourceFlags: %v", p.options.SourceFlags))
	p.lc.Info(fmt.Sprintf("IPRange: %v", p.options.IP))
	p.lc.Info(fmt.Sprintf("NetMask: %v", p.options.NetMask))
	p.lc.Info(fmt.Sprintf("Interval: %v", p.options.Interval))
	p.lc.Info(fmt.Sprintf("ScanDuration: %v", p.options.ScanDuration))
	p.lc.Info(fmt.Sprintf("Credentials (REMOVE): %v", p.options.Credentials))
	p.lc.Info(fmt.Sprintf("Cached ONVIF Devices: %v", len(p.ac.CamInfoCache.OnvifCameras)))
	p.lc.Info(fmt.Sprintf("Cached Axis Devices: %v", len(p.ac.CamInfoCache.AxisCameras)))

	// ==============================
	// Add service as EdgeX Device
	// ==============================
	labels := []string{"cameradiscovery"}
	deviceName := "camera-management-provider"
	err = p.registerDeviceManagementProvider(deviceName, labels)
	// ==============================
	// Schedule discovery scans
	// ==============================
	if err == nil {
		// Kick off our requested Discovery Schedule
		p.schedulePortScans(portList)
	}
	return err
}

// schedulePortScans declares a goroutine to initiate scans at requested interval
func (p *CameraDiscoveryProvider) schedulePortScans(portList string) {
	p.intervalTicker = time.NewTicker(time.Second * time.Duration(p.options.Interval))
	intervalCount := 0
	intervalStart := time.Now()
	go func() {
		for ; scanOnStartup; <-p.intervalTicker.C {
			intervalCount++
			deviceCount, err := p.DiscoverDevices(portList, *p.options)
			if err != nil {
				p.lc.Error(err.Error())
			}
			time.Sleep(1 * time.Second) // permits device sdk log entries to settle (affects presentation only)
			p.lc.Info(fmt.Sprintf("%v new camera devices registered during this scan", deviceCount))
			// Report next anticipated interval trigger
			nextScan := intervalStart.Local().Add(time.Second * time.Duration(p.options.Interval*intervalCount))
			remainSec := time.Until(nextScan)
			p.lc.Info(fmt.Sprintf("Next scan triggers in %.f seconds (%v), and each %v seconds thereafter.", remainSec.Seconds(), nextScan.Format(time.Stamp), p.options.Interval))
		}
	}()
}

func (p *CameraDiscoveryProvider) loadCameraInfoCaches() error {
	var fileOnvif string
	var fileAxis string
	if p.isVendorRequested(p.options.SourceFlags, "onvif") {
		fileOnvif = p.ac.InfoFileOnvif
	}
	if p.isVendorRequested(p.options.SourceFlags, "axis") {
		fileAxis = p.ac.InfoFileAxis
	}
	return p.ac.CamInfoCache.LoadInfo(p.lc, fileOnvif, fileAxis)
}

func (p *CameraDiscoveryProvider) getProtocols() map[string]contract.ProtocolProperties {
	p1 := make(map[string]string)
	p1["host"] = "localhost"
	p1["port"] = "all"

	p2 := make(map[string]string)
	p2["supports"] = "axis, onvif"

	wrap := make(map[string]contract.ProtocolProperties)
	wrap["connection"] = p1
	wrap["api_types"] = p2

	return wrap
}

func (p *CameraDiscoveryProvider) registerDeviceManagementProvider(deviceName string, labels []string) error {
	p.lc.Info(fmt.Sprintf("Adding CameraDeviceProvider as a proxy/manager EdgeX device"))
	// device.RunningService().RemoveDeviceByName(deviceName)
	edgexDevice, err := device.RunningService().GetDeviceByName(deviceName)
	if err != nil {

		idstr, err2 := device.RunningService().AddDevice(contract.Device{
			Name:           deviceName,
			AdminState:     contract.Unlocked,
			OperatingState: contract.Enabled,
			Protocols:      p.getProtocols(),
			Labels:   labels,
			Location: "gateway",
			Profile: contract.DeviceProfile{
				Name: CameraManagementProfileName,
			},
			Service: contract.DeviceService{
				AdminState:     contract.Unlocked,
				Name:           "device-camera-go",
				OperatingState: contract.Enabled,
			},
		})
		err = err2
		if err2 != nil {
			p.lc.Error("Error registering CameraDiscoveryProvider as EdgeX device: " + err2.Error())
		}
		// Upon success, edgex-core-metadata should also respond with a corresponding log message similar to:
		// INFO: 2018/11/14 18:30:46 AddDevice returned ID:5bec69d69f8fc20001fd3a6b
		time.Sleep(1 * time.Second)
		p.lc.Info("CameraDiscoveryProvider assigned EdgeX ID:" + idstr)
	} else {
		p.lc.Info(fmt.Sprintf("CameraDiscoveryProvider was previously registered and has EdgeX ID: %s", edgexDevice.Id))
	}
	return err
}

// HandleReadCommands passes a slice of CommandRequest struct each representing
// a ResourceOperation for a specific device resource.
func (p *CameraDiscoveryProvider) HandleReadCommands(deviceName string, protocols map[string]contract.ProtocolProperties, reqs []ds_models.CommandRequest) (res []*ds_models.CommandValue, err error) {
	if len(reqs) != 1 {
		err = fmt.Errorf("CameraDiscoveryDriver.HandleReadCommands; too many command requests; only one supported")
		return res, err
	}
	p.lc.Info(fmt.Sprintf("RECEIVED COMMAND REQUEST for DeviceResourceName %s  Attributes: [%v]", reqs[0].DeviceResourceName, reqs[0].Attributes))
	if reqs[0].DeviceResourceName == "onvif_profiles" || reqs[0].DeviceResourceName == "axis_info" {
		// These EdgeX commands are distinct for each device class.
		// ONVIF and vendor-specific APIs (e.g., Axis) provide different interfaces, often to the same physical device.
		var serialNum string
		var camInfo string
		if reqs[0].DeviceResourceName == "axis_info" {
			serialNum = strings.TrimPrefix(deviceName, p.options.SupportedSources[Axis].DeviceNamePrefix)
			camInfo = p.ac.CamInfoCache.TransformCameraInfoToString("axis", serialNum)
		}
		if reqs[0].DeviceResourceName == "onvif_profiles" {
			serialNum = strings.TrimPrefix(deviceName, p.options.SupportedSources[ONVIF].DeviceNamePrefix)
			camInfo = p.ac.CamInfoCache.TransformCameraInfoToString("onvif", serialNum)
		}
		res = make([]*ds_models.CommandValue, 1)
		now := time.Now().UnixNano() / int64(time.Millisecond)
		cv := ds_models.NewStringValue(reqs[0].DeviceResourceName, now, camInfo)
		res[0] = cv
	} else if reqs[0].DeviceResourceName == "tags" {
		// This EdgeX Command is common between two device classes (ONVIF and Axis)
		p.lc.Info(fmt.Sprintf("CameraDiscoveryProvider.HandleReadCommands: Returning Tags associated with device: %s", deviceName))
		serialNum := strings.TrimPrefix(deviceName, p.options.SupportedSources[ONVIF].DeviceNamePrefix)
		if len(serialNum) == len(deviceName) {
			serialNum = strings.TrimPrefix(deviceName, p.options.SupportedSources[Axis].DeviceNamePrefix)
		}
		camTags := createKeyValuePairString(p.ac.TagCache.Tags[serialNum])
		res = make([]*ds_models.CommandValue, 1)
		now := time.Now().UnixNano() / int64(time.Millisecond)
		cv := ds_models.NewStringValue(reqs[0].DeviceResourceName, now, camTags)
		res[0] = cv
	} else if reqs[0].DeviceResourceName == "get_user" {
		// Vendor specific command (Axis user CRUD example)
		p.lc.Info(fmt.Sprintf("CameraDiscoveryProvider.HandleReadCommands: TODO: Return EdgeX Video Users associated with device: %s", deviceName))
	}
	return
}

// HandleWriteCommands processes PUT commands, and is passed a slice of CommandRequest struct
// each representing a ResourceOperation for a specific device resource (aka DeviceObject).
// As these are actuation commands, params will provide parameters distinct to the command.
/*func (p *CameraDiscoveryProvider) HandleWriteCommands(addr *contract.Addressable, reqs []ds_models.CommandRequest,
	params []*ds_models.CommandValue) error {*/
func (p *CameraDiscoveryProvider) HandleWriteCommands(deviceName string, protocols map[string] contract.ProtocolProperties, reqs []ds_models.CommandRequest, params []*ds_models.CommandValue) error {
	if len(reqs) != 1 {
		err := fmt.Errorf("CameraDiscoveryDriver.HandleWriteCommands; too many command requests; only one supported")
		return err
	}
	p.lc.Info(fmt.Sprintf("TODO: CameraDiscoveryDriver.HandleWriteCommands: deviceName: %s protocolMap: [%v] reqs: [%v]", deviceName, protocols, reqs[0]))
	p.lc.Info(fmt.Sprintf("with params: %v", params))
	if reqs[0].Attributes["name"] == "tags" {
		p.lc.Info(fmt.Sprintf("CameraDiscoveryProvider.HandleWriteCommands: TODO: PUT tags caller wants associated with device: %s", deviceName))
	} else if reqs[0].Attributes["name"] == "user" {
		// TODO: To support CRUD add commands for /add_user, /update_user, /remove_user
		p.lc.Info(fmt.Sprintf("CameraDiscoveryProvider.HandleWriteCommands: TODO: PUT user group and credentials for EdgeX Video User that caller wants associated with device: %s", deviceName))
	}
	return nil
}

//Stop is called on termination of service. Perform any needed cleanup for graceful/forced shutdown here.
func (p *CameraDiscoveryProvider) Stop(force bool) error {
	p.lc.Debug("Stopping intervalTicker")
	p.intervalTicker.Stop()

	p.lc.Debug(fmt.Sprintf("Stop Called: force=%v", force))
	return nil
}
